File: lib/services/auth_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:firebase_auth/firebase_auth.dart' as fb_auth;
import '../models/user.dart';

class AuthService {
  static const String _userKey = 'user';
  static const String _tokenKey = 'token';

  final SharedPreferences _prefs;

  AuthService(this._prefs);

  final fb_auth.FirebaseAuth _firebaseAuth = fb_auth.FirebaseAuth.instance;

  // Public getter for FirebaseAuth instance
  fb_auth.FirebaseAuth get firebaseAuthInstance => _firebaseAuth;

  static Future<AuthService> create() async {
    final prefs = await SharedPreferences.getInstance();
    return AuthService(prefs);
  }

  Future<bool> isLoggedIn() async {
    // Primarily check Firebase auth state
    return _firebaseAuth.currentUser != null;
    // We might also want to check _prefs.containsKey(_tokenKey) depending on session management strategy
  }

  Future<void> saveUserSession(User user, String token) async {
    await _prefs.setString(_userKey, jsonEncode(user.toJson()));
    await _prefs.setString(_tokenKey, token);
  }

  Future<void> logout() async {
    await _firebaseAuth.signOut();
    await _prefs.remove(_userKey);
    await _prefs.remove(_tokenKey);
  }

  Future<User?> getCurrentUser() async {
    final userJson = _prefs.getString(_userKey);
    if (userJson == null) return null;
    return User.fromJson(jsonDecode(userJson));
  }

  Future<String?> getToken() async {
    return _prefs.getString(_tokenKey);
  }

  // Firebase Phone Authentication Methods

  fb_auth.User? get currentFirebaseUser => _firebaseAuth.currentUser;

  Future<void> verifyPhoneNumber({
    required String phoneNumber,
    required Function(fb_auth.PhoneAuthCredential) verificationCompleted,
    required Function(fb_auth.FirebaseAuthException) verificationFailed,
    required Function(String, int?) codeSent,
    required Function(String) codeAutoRetrievalTimeout,
    int? forceResendingToken, // Added this line
    Duration timeout = const Duration(seconds: 60),
  }) async {
    await _firebaseAuth.verifyPhoneNumber(
      phoneNumber: phoneNumber,
      verificationCompleted: verificationCompleted,
      verificationFailed: verificationFailed,
      codeSent: codeSent,
      codeAutoRetrievalTimeout: codeAutoRetrievalTimeout,
      forceResendingToken: forceResendingToken, // Added this line
      timeout: timeout,
    );
  }

  Future<fb_auth.UserCredential?> signInWithOtp(String verificationId, String smsCode) async {
    try {
      final fb_auth.PhoneAuthCredential credential = fb_auth.PhoneAuthProvider.credential(
        verificationId: verificationId,
        smsCode: smsCode,
      );
      return await _firebaseAuth.signInWithCredential(credential);
    } catch (e) {
      // Handle error, e.g., invalid OTP
      print('Error signing in with OTP: $e');
      return null;
    }
  }

  Future<fb_auth.UserCredential?> signInWithPhoneCredential(fb_auth.PhoneAuthCredential credential) async {
    try {
      return await _firebaseAuth.signInWithCredential(credential);
    } catch (e) {
      print('Error signing in with phone credential: $e');
      return null;
    }
  }

  // Method to link phone number to an existing anonymous or other user (optional)
  Future<void> linkPhoneNumber(String verificationId, String smsCode) async {
    final credential = fb_auth.PhoneAuthProvider.credential(
      verificationId: verificationId,
      smsCode: smsCode,
    );
    await _firebaseAuth.currentUser?.linkWithCredential(credential);
  }

  // Method to update phone number for an existing user (optional)
  // This often requires recent re-authentication.
  Future<void> updatePhoneNumber(String verificationId, String smsCode) async {
     final credential = fb_auth.PhoneAuthProvider.credential(
      verificationId: verificationId,
      smsCode: smsCode,
    );
    await _firebaseAuth.currentUser?.updatePhoneNumber(credential);
  }

  // Firebase Email/Password Authentication

  Future<fb_auth.User?> createUserWithEmailAndPasswordAndName(String email, String password, String name) async {
    try {
      final fb_auth.UserCredential userCredential = await _firebaseAuth.createUserWithEmailAndPassword(
        email: email,
        password: password,
      );
      // Update the user's display name
      await userCredential.user?.updateDisplayName(name);
      // Reload the user to get the updated display name
      await userCredential.user?.reload();
      return _firebaseAuth.currentUser; // Return the user with updated info
    } on fb_auth.FirebaseAuthException catch (e) {
      print('FirebaseAuthException during email registration: ${e.code} - ${e.message}');
      rethrow;
    } catch (e) {
      print('Error creating user with email and password: $e');
      return null;
    }
  }

  Future<fb_auth.User?> signInWithEmailAndPassword(String email, String password) async {
    try {
      final fb_auth.UserCredential userCredential = await _firebaseAuth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );
      return userCredential.user;
    } on fb_auth.FirebaseAuthException catch (e) {
      // Specific Firebase Auth errors can be handled or re-thrown here
      // For example, you might want to log them or map them to custom error types
      print('FirebaseAuthException during email sign-in: ${e.code} - ${e.message}');
      rethrow; // Re-throw to be caught by AuthProvider or UI layer
    } catch (e) {
      print('Error signing in with email and password: $e');
      // Handle other errors, e.g., network issues
      return null; // Or rethrow a generic exception
    }
  }
}

File: src/services/authService.ts
import {
    signInWithEmailAndPassword as firebaseSignInWithEmail,
    createUserWithEmailAndPassword,
    signInWithPhoneNumber,
    signInWithCredential,
    RecaptchaVerifier,
    ConfirmationResult,
    updateProfile,
    signOut,
    User as FirebaseUser,
    PhoneAuthCredential,
    Auth,
} from 'firebase/auth';
import { auth, isDemoMode } from './firebase';
import { User } from '@/types';

const USER_KEY = 'elite_customer_user';
const TOKEN_KEY = 'elite_customer_token';

// Initialize reCAPTCHA verifier for phone auth
let recaptchaVerifier: RecaptchaVerifier | null = null;

export const setupRecaptcha = (containerId: string): RecaptchaVerifier => {
    if (typeof window !== 'undefined' && auth) {
        recaptchaVerifier = new RecaptchaVerifier(auth, containerId, {
            size: 'invisible',
            callback: () => {
                // reCAPTCHA solved
            },
            'expired-callback': () => {
                // Reset reCAPTCHA
                recaptchaVerifier = null;
            },
        });
    }
    return recaptchaVerifier!;
};

export const getRecaptchaVerifier = (): RecaptchaVerifier | null => recaptchaVerifier;

// Check if user is logged in
export const isLoggedIn = (): boolean => {
    if (isDemoMode || !auth) {
        return getStoredUser() !== null;
    }
    return auth.currentUser !== null;
};

// Get current Firebase user
export const getCurrentFirebaseUser = (): FirebaseUser | null => {
    if (!auth) return null;
    return auth.currentUser;
};

// Save user session to localStorage
export const saveUserSession = (user: User, token: string): void => {
    if (typeof window !== 'undefined') {
        localStorage.setItem(USER_KEY, JSON.stringify(user));
        localStorage.setItem(TOKEN_KEY, token);
    }
};

// Get stored user from localStorage
export const getStoredUser = (): User | null => {
    if (typeof window === 'undefined') return null;
    const userJson = localStorage.getItem(USER_KEY);
    if (!userJson) return null;
    try {
        return JSON.parse(userJson) as User;
    } catch {
        return null;
    }
};

// Get stored token from localStorage
export const getStoredToken = (): string | null => {
    if (typeof window === 'undefined') return null;
    return localStorage.getItem(TOKEN_KEY);
};

// Clear stored session
export const clearSession = (): void => {
    if (typeof window !== 'undefined') {
        localStorage.removeItem(USER_KEY);
        localStorage.removeItem(TOKEN_KEY);
    }
};

// Logout
export const logout = async (): Promise<void> => {
    if (auth) {
        await signOut(auth);
    }
    clearSession();
};

// Email/Password Authentication
export const signInWithEmailAndPassword = async (
    email: string,
    password: string
): Promise<FirebaseUser> => {
    if (!auth) {
        throw new Error('Firebase auth not initialized. Running in demo mode.');
    }
    const userCredential = await firebaseSignInWithEmail(auth, email, password);
    return userCredential.user;
};

export const createUserWithEmailAndPasswordAndName = async (
    email: string,
    password: string,
    name: string
): Promise<FirebaseUser> => {
    if (!auth) {
        throw new Error('Firebase auth not initialized. Running in demo mode.');
    }
    const userCredential = await createUserWithEmailAndPassword(auth, email, password);
    await updateProfile(userCredential.user, { displayName: name });
    return userCredential.user;
};

// Phone Authentication
export const verifyPhoneNumber = async (
    phoneNumber: string,
    recaptcha: RecaptchaVerifier
): Promise<ConfirmationResult> => {
    if (!auth) {
        throw new Error('Firebase auth not initialized. Running in demo mode.');
    }
    return signInWithPhoneNumber(auth, phoneNumber, recaptcha);
};

export const confirmOtp = async (
    confirmationResult: ConfirmationResult,
    otp: string
): Promise<FirebaseUser> => {
    const userCredential = await confirmationResult.confirm(otp);
    return userCredential.user;
};

export const signInWithPhoneCredential = async (
    credential: PhoneAuthCredential
): Promise<FirebaseUser> => {
    if (!auth) {
        throw new Error('Firebase auth not initialized. Running in demo mode.');
    }
    const userCredential = await signInWithCredential(auth, credential);
    return userCredential.user;
};

// Create User object from Firebase user
export const createUserFromFirebase = async (firebaseUser: FirebaseUser): Promise<User> => {
    const token = await firebaseUser.getIdToken();
    const user: User = {
        id: firebaseUser.uid,
        name: firebaseUser.displayName || 'User',
        email: firebaseUser.email || '',
        phone: firebaseUser.phoneNumber || '',
    };
    saveUserSession(user, token);
    return user;
};

// Update user display name
export const updateUserDisplayName = async (name: string): Promise<void> => {
    if (!auth) return;
    const user = auth.currentUser;
    if (user) {
        await updateProfile(user, { displayName: name });
    }
};

// Demo mode login - bypasses Firebase for testing
export const demoLogin = (): User => {
    const demoUser: User = {
        id: 'demo-user-123',
        name: 'Demo User',
        email: 'demo@elite.com',
        phone: '+91 9876543210',
    };
    saveUserSession(demoUser, 'demo-token');
    return demoUser;
};

export default {
    isLoggedIn,
    getCurrentFirebaseUser,
    saveUserSession,
    getStoredUser,
    getStoredToken,
    clearSession,
    logout,
    signInWithEmailAndPassword,
    createUserWithEmailAndPasswordAndName,
    verifyPhoneNumber,
    confirmOtp,
    signInWithPhoneCredential,
    createUserFromFirebase,
    updateUserDisplayName,
    setupRecaptcha,
    getRecaptchaVerifier,
    demoLogin,
    isDemoMode,
};
